server/src/test/java/io/crate/planner/operators/JoinTest.java

```diff
@@ -1241,4 +1243,38 @@ public void test_many_mized_implicit_and_explicit_joins() throws Exception {
             "  └ Collect[doc.t6 | [f] | true]"
         );
     }
+
+    @Test
+    public void test_hash_join_on_simple_pk_column_generate_hash_join_phase_with_isUniqueHashPerRow_set() throws IOException {
+        e = SQLExecutor.of(clusterService)
+            .addTable("create table t1 (a int primary key)")
+            .addTable("create table t2 (a int primary key, b int primary key)")
+            .addTable("create table t3 (a int, b int)")
+            .addTable("create table t4 (a int primary key, b int)");
+
+        QueriedSelectRelation analyzed = e.analyze("select * from t1 join t3 on t1.a = t3.a");
+        JoinPhase joinPhase = buildJoin(buildLogicalPlan(analyzed)).joinPhase();
+        assertThat(joinPhase).isExactlyInstanceOf(HashJoinPhase.class);
+        assertThat(((HashJoinPhase) joinPhase).isUniqueHashPerRow()).isTrue();
+
+        // column 't2.a' alone cannot uniquely identify each row
+        analyzed = e.analyze("select * from t2 join t3 on t2.a = t3.a");
+        joinPhase = buildJoin(buildLogicalPlan(analyzed)).joinPhase();
+        assertThat(joinPhase).isExactlyInstanceOf(HashJoinPhase.class);
+        assertThat(((HashJoinPhase) joinPhase).isUniqueHashPerRow()).isFalse();
+
+        // columns 't2.a' and 't3.a' together uniquely identify each row, but this does not guarantee
+        // that their combined hash value is unique; hence hash collisions are possible
+        analyzed = e.analyze("select * from t2 join t3 on t2.a = t3.a and t2.b = t3.b");
+        joinPhase = buildJoin(buildLogicalPlan(analyzed)).joinPhase();
+        assertThat(joinPhase).isExactlyInstanceOf(HashJoinPhase.class);
+        assertThat(((HashJoinPhase) joinPhase).isUniqueHashPerRow()).isFalse();
+
+        // column 't4.a' uniquely identify each row, but the combined hash value with 't4.b' cannot be unique anymore
+        // (hash collisions are possible with the combined hash)
+        analyzed = e.analyze("select * from t4 join t3 on t4.a = t3.a and t4.b = t3.b");
+        joinPhase = buildJoin(buildLogicalPlan(analyzed)).joinPhase();
+        assertThat(joinPhase).isExactlyInstanceOf(HashJoinPhase.class);
+        assertThat(((HashJoinPhase) joinPhase).isUniqueHashPerRow()).isFalse();
```

Comment:
We can push a follow-up change to hash only on the primary key column. This would allow joins involving a PK column and an additional non-PK column(i.e. `on t4.a = t3.a and t4.b = t3.b`) to achieve similar performance improvements.

Currently, in the example, LHS rows are hashed using both `t4.a` and `t4.b`, and RHS rows are matched based on the hash of `t3.a` and `t3.b`.

Instead, we could hash LHS rows using only `t4.a` and match RHS rows using only `t3.a`.